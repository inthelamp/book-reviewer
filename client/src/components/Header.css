#user-nav-dropdown {
    width: 8vw !important;
    text-align: right;
}