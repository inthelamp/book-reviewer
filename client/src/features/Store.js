import { createStore, combineReducers } from 'redux'
import UserReducer from './UserReducer'
import TextEditorReducer from './TextEditorReducer'

// Creating store
const rootReducer = combineReducers({ User: UserReducer, TextEditor: TextEditorReducer})
const Store = createStore(
        rootReducer,
        window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__() 
)

export default Store