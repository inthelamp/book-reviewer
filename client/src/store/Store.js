import { createStore, combineReducers } from 'redux'
import UserReducer from './UserReducer'


//const rootReducer = combineReducers({User: UserReducer})
const Store = createStore(UserReducer)

export default Store