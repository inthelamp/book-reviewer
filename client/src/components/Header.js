import { useMemo } from 'react'
import { useSelector } from 'react-redux'
import { Link, useHistory } from 'react-router-dom'
import 'bootstrap/dist/css/bootstrap.min.css'
import {Form, FormControl, Button, Nav, Navbar, NavDropdown} from 'react-bootstrap'
import PropTypes from 'prop-types'
import { persistentUser } from '../features/PersistentUser'
import './Header.css'


const Header = ({ title }) => {
    const storeUser = useSelector((state) => state)
    const User = Object.keys(storeUser).length === 0 ? persistentUser() : storeUser

    const getUserName = (user) => {
        let name = '' 

        if (user && user.isSignedIn) {            
            const role = user.role        
            name = user.name
    
            if (role === 'Admin') {
                name = name + ' (' + role + ')'
            }
        }

        return name        
    }    

    // attach Sign Out tab to user name if user name is available
    const userName = useMemo(() => getUserName(User), [User])

    const userTab = userName === '' ? 
                (        
                    <Nav.Link href='/signin'>Sign In</Nav.Link>
                )
                : 
                (
                    <NavDropdown title={userName} id='user-nav-dropdown'>
                        <NavDropdown.Item href='/signout'>
                            Sign Out
                        </NavDropdown.Item>
                    </NavDropdown>
                )     

    // sign out when token is expired            
    const history = useHistory()            

    const signOut = () => {
        history.push('/signout')
    }            

    if (User && User.isTokenExpried) {
        signOut()
    } 
                                  
    return (
        <Navbar bg='white' expand='lg'>            
            <Navbar.Brand 
                    as={Link}  
                    to={{
                        pathname: '/',
                        state: { User }
                    }}
            >
                { title }
            </Navbar.Brand>
            <Navbar.Toggle aria-controls='basic-navbar-nav' />
            <Navbar.Collapse id='basic-navbar-nav'>
            <Nav className='me-auto'>
                <Nav.Link
                    as={Link} 
                    to={{
                        pathname: '/browse',
                        state: { User }
                    }}                
                >
                    Read Reviews
                </Nav.Link>
                <Nav.Link 
                    as={Link} 
                    to={{
                        pathname: '/write',
                        state: { User }
                    }}
                >
                    Write Review
                </Nav.Link>
                <Form className='d-flex'>
                    <FormControl
                        type='search'
                        placeholder='Search'
                        className='mt-0 mb-0 mr-02'
                        aria-label='Search'
                    />
                    <Button variant='success' size='sm'>Search</Button>
                </Form>                  
            </Nav>          
            <Nav className='pull-right user-tab'>
                { userTab }
            </Nav>                 
            </Navbar.Collapse>
        </Navbar>        
    )
}

Header.defaultProps = {
    title: 'Book Reviewer',
}

Header.propTypes = {
    title: PropTypes.string.isRequired,
}

export default Header
