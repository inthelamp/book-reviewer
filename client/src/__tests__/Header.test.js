import { BrowserRouter as Router } from 'react-router-dom'
import renderer from 'react-test-renderer'
import { Provider } from 'react-redux'
import Store from '../features/Store'
import Header from "../components/Header"

it('Component: Header rendering correctly', () => {  
    const tree = renderer
      .create(<Provider store={Store}><Router><Header /></Router></Provider>)
      .toJSON();
    expect(tree).toMatchSnapshot()
})