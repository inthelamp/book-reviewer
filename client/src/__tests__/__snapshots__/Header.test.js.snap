// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Header rendering correctly 1`] = `
<header>
  <nav
    className="navbar navbar-expand-lg navbar-light bg-white"
  >
    <img
      alt="Logo"
      className="App-logo"
      src={Object {}}
    />
    <a
      className="navbar-brand"
      href="/"
      onClick={[Function]}
    >
      Book Reviewer
    </a>
    <div
      aria-expanded={null}
      className="navbar-collapse collapse"
      id="basic-navbar-nav"
    >
      <div
        className="me-auto navbar-nav"
        onKeyDown={[Function]}
      />
      <div
        className="pull-right user-tab navbar-nav"
        onKeyDown={[Function]}
      >
        <a
          className="nav-link"
          data-rr-ui-event-key="/signin"
          disabled={false}
          href="/signin"
          onClick={[Function]}
        >
          Sign In
        </a>
      </div>
    </div>
  </nav>
</header>
`;
